sql/src/main/java/io/crate/operation/collect/collectors/ShardStateAwareRemoteCollector.java

```diff
@@ -105,86 +102,101 @@ public void doCollect() {
+                        } else {
+                            consumer.accept(null, new InterruptedException());
                             return;
                         }
-                        try {
-                            synchronized (killLock) {
-                                if (collectorKilled == false) {
-                                    IndexShard indexShard = indexShards.getShard(shardId);
-                                    collector = localCollectorProvider.apply(indexShard);
-                                } else {
-                                    consumer.accept(null, new InterruptedException());
-                                    return;
-                                }
-                            }
-                            // collect outside the synchronized block as the collectors handle kill signals that arrive
-                            // while collecting
-                            if (collector != null) {
-                                collector.doCollect();
-                            }
-                        } catch (Exception e) {
-                            consumer.accept(null, e);
-                        }
-                    } else {
-                        consumer.accept(null, new TableUnknownException(TableIdent.fromIndexName(indexName)));
                     }
-                });
+                    // collect outside the synchronized block as the collectors handle kill signals that arrive
+                    // while collecting
+                    if (collector != null) {
+                        collector.doCollect();
+                    }
+                } catch (Exception e) {
+                    consumer.accept(null, e);
+                }
             } else {
-                LOGGER.warn("Unable to collect from a shard that's not started. Waiting for primary shard {} to start.",
-                    shardRouting.shardId());
-
-                ClusterStateObserver clusterStateObserver = new ClusterStateObserver(clusterState, clusterService,
-                    stateObserverWaitForChangeTimeout, LOGGER, threadContext);
-                clusterStateObserver.waitForNextChange(getClusterStateListener(shardRouting),
-                    newState -> {
-                        try {
-                            return newState.routingTable().shardRoutingTable(indexName, shardId).primaryShard().started();
-                        } catch (IndexNotFoundException | ShardNotFoundException e) {
-                            // we'll handle this as a shard/index deleted change
-                            return true;
-                        }
-                    });
+                consumer.accept(null, new IndexNotFoundException(indexName));
             }
-        }
+        });
+    }
+
+
+    private void addClusterStateObserverWaitingForNewShardState(ClusterState clusterState) {
+        LOGGER.warn("Waiting for primary shard {} to be started.", shardId);
```

Comment:
maybe wrap inside a `isWarnEnabled` block to avoid varargs object array allocations